src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs

```
@@ -223,5 +237,44 @@ private static IList<string> GetOrderBy(V2SortBy sortBy)
 
             return orderBy;
         }
+
+        // Constructs filter strings for both Frameworks and Tfms.
+        // indexField: Determines which field you are targeting
+        //             i.e. IndexFields.Search.Frameworks or IndexFields.Search.Tfms
+        // frameworks: Comma-separated list of a user's selected Frameworks or Tfms
+        private string GetFrameworksOrTfmsFilterString(string indexField, IReadOnlyList<string> frameworks)
+        {
+            var filterStrings = frameworks
+                                    .Where(f => IsValidFrameworkOrTfm(indexField, f))
+                                    .Select(f => $"{indexField}/any(f: f eq '{f.ToLowerInvariant()}')");
+
+            return filterStrings.Count() == 0
+                                ? String.Empty
+                                : " and (" + String.Join(" and ", filterStrings) + ")";
+        }
+
+        private bool IsValidFrameworkOrTfm(string indexField, string framework)
```

For the `Tfms` flow, we're using the original (user provided) value not the normalized version. Consider changing `IsValidFrameworkOrTfm` to `NormalizeOrNull` method that normalizes the input and returns null on invalid, Then you can just filter null after calling it.

---

BTW, the behavior we have here is to _ignore_ bad input instead of erroring. We could return an error to the user if the input is not valid instead of ignoring it silently.

In general search service ignores bad input which ... maybe isn't a great dev experience? We could opt to be more explicit here.

Thoughts?

---

I've moved all the validating and normalizing code over to the Controller layer (ParameterUtilities.cs), and it looks a bit different now, so I'm not sure how the first comment applies. We are still ignoring invalid inputs rather than erroring with the current code though. We can flag invalid framework or tfm strings when we're validating them in the Controller. Do we have other examples of invalid parameter errors in Search? I'd like to see what the UI experience (and the code) for this might look like

---

No examples right now. In general we don't error out on bad input, we ignore it. And the error would probably not be surfaced in the Gallery UI at all. It would be more for custom code against our APIs (3rd party devs)